pdns/bpf-filter.cc

```diff
@@ -470,6 +506,101 @@ void BPFFilter::unblock(const ComboAddress& addr)
   }
 }
 
+void BPFFilter::block(const Netmask& addr, BPFFilter::MatchAction action)
+{
+  CounterAndActionValue value;
+
+  int res = 0;
+  if (addr.isIPv4()) {
+    CIDR4 key(addr);
+    auto maps = d_maps.lock();
+    auto& map = maps->d_cidr4;
+    if (map.d_count >= map.d_config.d_maxItems) {
+      throw std::runtime_error("Table full when trying to block " + addr.toString());
+    }
+
+    res = bpf_lookup_elem(map.d_fd.getHandle(), &key, &value);
+    if (res != -1 && value.action == action) {
+      throw std::runtime_error("Trying to block an already blocked address: " + addr.toString());
+    }
+
+    value.counter = 0;
+    value.action = action;
+
+    res = bpf_update_elem(map.d_fd.getHandle(), &key, &value, BPF_NOEXIST);
+    if (res == 0) {
+      ++map.d_count;
+    }
+  }
+  else if (addr.isIPv6()) {
+    CIDR6 key(addr);
+
+    auto maps = d_maps.lock();
+    auto& map = maps->d_cidr6;
+    if (map.d_count >= map.d_config.d_maxItems) {
+      throw std::runtime_error("Table full when trying to block " + addr.toString());
+    }
+
+    res = bpf_lookup_elem(map.d_fd.getHandle(), &key, &value);
+    if (res != -1 && value.action == action) {
+      throw std::runtime_error("Trying to block an already blocked address: " + addr.toString());
+    }
+
+    value.counter = 0;
+    value.action = action;
+
+    res = bpf_update_elem(map.d_fd.getHandle(), &key, &value, BPF_NOEXIST);
+    if (res == 0) {
+      map.d_count++;
+    }
+  }
+
+  if (res != 0) {
+    throw std::runtime_error("Error adding blocked address " + addr.toString() + ": " + stringerror());
+  }
+}
+
+void BPFFilter::unblock(const Netmask& addr)
+{
+  int res = 0;
+  CounterAndActionValue value;
+  value.counter = 0;
+  value.action = MatchAction::Pass;
+  if (addr.isIPv4()) {
+    CIDR4 key(addr);
+    auto maps = d_maps.lock();
+    auto& map = maps->d_cidr4;
+    res = bpf_delete_elem(map.d_fd.getHandle(), &key);
+    if (res == 0) {
+      --map.d_count;
+    }
+    else {
+      res = bpf_update_elem(map.d_fd.getHandle(), &key, &value, BPF_NOEXIST);
```

Comment:
I'm likely missing something here, why are we inserting if the deletion failed? Or is that intended as a way to allow-list a more specific entry, hence the `MatchAction::Pass` action? But then we would unblock() again to remove the allow-listed entry, which seems a bit counter-intuitive?

Reply:
If we want to block an ip range, but also give an exception.
We might do like this:

```
bpf:blockRange("192.68.1.0/24")
bpf:unblockRange("192.168.1.0/28")`
```

When we unblock `192.168.1.0/28`, `bpf_delete_elem` could not find a item with key `192.168.1.0/28`. So,we need add a item with key `192.168.1.0/28` and value `MatchAction::Pass`.

I admit it's a bit counter-intuitive. But I can't think of a better way to do it.

Reply:
Yes, that's clever. Perhaps we could rename `unblockRange` to `allowRange`, or something like that, and document that it will either remove any existing entry blocking that range, or add a new entry explicitly allowing that range otherwise?

Reply:
Oh yeah, renaming `unblockRange` seems like a good way to describe this feature more clearly.